lib/plugins/aws/deployFunction.js

```diff
@@ -103,10 +103,27 @@ class AwsDeployFunction {
       FunctionName: functionObj.name,
     };
 
-    if ('awsKmsKeyArn' in functionObj && !_.isObject(functionObj.awsKmsKeyArn)) {
-      params.KMSKeyArn = functionObj.awsKmsKeyArn;
-    } else if (serviceObj && 'awsKmsKeyArn' in serviceObj && !_.isObject(serviceObj.awsKmsKeyArn)) {
-      params.KMSKeyArn = serviceObj.awsKmsKeyArn;
+    let kmsKeyArn = functionObj.kmsKeyArn || providerObj.kmsKeyArn;
+    if (functionObj.awsKmsKeyArn) {
```

Comment:
What is your opinion on moving the deprecation checks to `deploy:function:initialize` hook? We use similar approach in https://github.com/serverless/serverless/blob/master/lib/plugins/aws/package/compile/functions.js#L33-L63 which I believe helps with readability and nicely separates deprecation checks with the actual logic

Reply:
I like it. I don't love the `if` checks:
```javascript
if (_.get(this.serverless.service.serviceObject, 'awsKmsKeyArn'))

if (Object.values(this.serverless.service.functions).some(({ awsKmsKeyArn }) => awsKmsKeyArn))
```

Is there a reason I couldn't do this? Could the `awsKmsKeyArn` live anywhere other than the `functionObj` property within the `serverless.service.functions`.
```diff
- if (_.get(this.serverless.service.serviceObject, 'awsKmsKeyArn'))
+ if (this.serverless.service.serviceObject.awsKmsKeyArn)

- if (Object.values(this.serverless.service.functions).some(({ awsKmsKeyArn }) => awsKmsKeyArn))
+ if (this.serverless.service.functions.functionObj.awsKmsKeyArn)
```
And then as a general question, when exactly does that hook get run?

Reply:
It looks good. A couple of questions:

Can I modify the if checks?
```diff
- if (_.get(this.serverless.service.serviceObject, 'awsKmsKeyArn'))
+ if (this.serverless.service.serviceObject.awsKmsKeyArn)
 
- if (Object.values(this.serverless.service.functions).some(({ awsKmsKeyArn }) => awsKmsKeyArn))
+ if (this.serverless.service.functions.functionObj.awsKmsKeyArn)
```
Will the `awsKmsKeyArn` property ever exist on anything other than the `functionObj`? If so, I understand the use of the `.some()` method. Otherwise, we can probably just check `functionObj` for that property directly. I guess unless `functionObj` is not guaranteed to exist.

And then just a general question -- when does the hook run?

Reply:
As for when do hooks run - they're invoked as a part of `deploy` command, you can see the logic here: https://github.com/serverless/serverless/blob/b093609f7952d5a63c91e6435b6a3a7d7d09cb1a/lib/plugins/deploy.js#L58

As for checks:
- the use of `some` is there due to the fact that in `functions.js` we're considering all functions - here it will be enough if we check only on the one that is considered for deploy
- the check for `awsKmsKeyArn` on `serviceObject` - here I'm not 100% if there's a use case where where the `serviceObject` could be empty (I'm not aware of it), but for now I would stick to `_.get` use - we have a separate issue where we migrate `lodash` use and we would be able to address it at a later point in time